fix: dark mode styling for trace view with design tokens

jkowall opened the pull request:
## Summary

This PR fixes multiple dark mode rendering issues in the trace view and establishes proper CSS design token usage for theming support.

## Screenshots

**Before (issues):**
- Minimap showing white background
- Expand/collapse arrows invisible
- Timeline header unreadable
- Back button not visible
- Theme toggle button showing black
- Search input text black

**After:** All elements properly themed with correct contrast and visibility.

## Root Cause Discovery: Canvas Background Hardcoded in JavaScript

The most significant finding was that the **minimap was impossible to fix with CSS alone**. The canvas background was hardcoded in JavaScript:

```typescript
// render-into-canvas.tsx (before)
export const BG_COLOR = '#fff';
// ...
ctx.fillStyle = BG_COLOR;
ctx.fillRect(0, 0, cWidth, cHeight);
```

This explains why PR #3244 and other CSS-only approaches could not fix the minimap. The solution required:
1. Adding theme detection: `document.body.getAttribute('data-theme')`
2. Supporting both light (`#fff`) and dark (`#0b1625`) backgrounds
3. Adding `useThemeMode()` hook to trigger re-render on theme changes

## Issues Fixed

### 1. Minimap (SpanGraph) Canvas Background
- **Root cause:** JavaScript hardcoded `#fff` background
- **Fix:** Theme-aware `getBackgroundColor()` function + re-render on theme change

### 2. Timeline Header Row (`position: fixed`)
- **Issue:** CSS variables scoped to `.ant-layout` not available for fixed elements
- **Fix:** Added `body` selector alongside `.ant-layout` in vars.css

### 3. Expand/Collapse Arrows (SpanTreeOffset)
- **Issue:** `color: #000` invisible on dark backgrounds
- **Fix:** Uses `--span-tree-icon-color` token

### 4. Back Button & Trace Overview (TracePageHeader)
- **Issue:** Hardcoded grays (`#fafafa`, `#eee`, `#ddd`)
- **Fix:** Uses `--surface-*` and `--border-*` tokens

### 5. Top Navigation Elements
- **Issue:** `ThemeToggleButton` used `--text-inverse` (black in dark mode)
- **Issue:** `TraceIDSearchInput` had no dark mode styling
- **Fix:** Explicit light colors for always-dark nav header

### 6. Search Bar & Keyboard Shortcuts
- **Issue:** Hardcoded colors in `TracePageSearchBar` and `KeyboardShortcutsHelp`
- **Fix:** Uses design tokens

## Design Token Architecture

### Token Definition Strategy
```css
body,
.ant-layout {
  /* Fallbacks for when --ant-* not available */
  --surface-primary: var(--ant-color-bg-container, #ffffff);
  --border-default: var(--ant-color-border-secondary, #f0f0f0);
  /* ... */
}

body[data-theme='dark'],
[data-theme='dark'] .ant-layout {
  /* Explicit dark mode values */
  --surface-primary: #162338;
  --border-default: rgba(125, 153, 191, 0.25);
  /* ... */
}
```

### Why Both `body` and `.ant-layout`?
- `.ant-layout` - For components inside Ant Design Layout
- `body` - For `position: fixed` elements and components outside Layout

### Fallback Values
All tokens include fallbacks for when `--ant-*` variables aren't available:
```css
--surface-primary: var(--ant-color-bg-container, #ffffff);
```

## New Design Tokens

| Category | Tokens |
|----------|--------|
| Timeline | `--timeline-header-bg`, `--timeline-tick-color`, `--timeline-tick-label-color` |
| Span Tree | `--span-tree-icon-color`, `--span-tree-hover-bg`, `--span-tree-indent-guide`, `--span-tree-indent-guide-active` |
| Minimap | `--span-graph-bg`, `--span-graph-border`, `--span-graph-inactive` |
| Scrubber | `--scrubber-handle-color`, `--scrubber-active-color` |
| Collapser | `--timeline-collapser-color`, `--timeline-collapser-hover` |
| Graph | `--graph-tick-stroke` |

## Files Changed

### Core Changes
- `vars.css` - Centralized design token definitions with dark mode overrides
- `render-into-canvas.tsx` - Theme-aware canvas background color
- `CanvasSpanGraph.tsx` - Re-render on theme change via `useThemeMode()` hook

### CSS Files Updated (replaced hardcoded colors with tokens)
- `SpanGraph/CanvasSpanGraph.css`
- `SpanGraph/GraphTicks.css`
- `SpanGraph/Scrubber.css`
- `SpanGraph/TickLabels.css`
- `SpanGraph/ViewingLayer.css`
- `SpanGraph/SpanGraph.css` (new file)
- `TracePageHeader/TracePageHeader.css`
- `TracePageHeader/TracePageSearchBar.css`
- `TracePageHeader/AltViewOptions.css`
- `TracePageHeader/KeyboardShortcutsHelp.css`
- `TimelineHeaderRow/TimelineHeaderRow.css`
- `TimelineHeaderRow/TimelineCollapser.css`
- `TraceTimelineViewer/SpanTreeOffset.css`
- `TraceTimelineViewer/Ticks.css`
- `App/ThemeToggleButton.css`
- `App/TraceIDSearchInput.css`

### Test Files Updated
- `render-into-canvas.test.js` - Updated export name `BG_COLOR` → `BG_COLOR_LIGHT`
- `CanvasSpanGraph.test.js` - Added mock for `useThemeMode()`

## Testing

- [x] All existing tests pass
- [x] TypeScript type checking passes
- [x] Light mode appearance unchanged
- [x] Dark mode all trace view elements visible and properly themed
- [x] Theme toggle correctly updates minimap canvas

## Related

- Addresses similar issues as PR #3244 but uses design tokens instead of hardcoded colors
- Follows the theming architecture from ADR-0001

yurishkuro commented on packages/jaeger-ui/src/components/common/vars.css:
```diff
@@ -8,18 +8,25 @@ SPDX-License-Identifier: Apache-2.0
+ * for fixed-position elements and components rendered outside the Layout.
+ * 
+ * Light mode uses fallback values since --ant-* variables may not be available
+ * on body element. Dark mode provides explicit overrides.
```
yes, this was quite a fight yesterday to get working, but sounds like it still has gaps. What are the examples of "fixed-position elements and components rendered outside the Layout"?

yurishkuro replied:
I think #3247 fixes that issue by making both `--ant-xxx` and our custom variables visible at the `:root` level.

yurishkuro commented on packages/jaeger-ui/src/components/common/vars.css:
```diff
@@ -8,18 +8,25 @@ SPDX-License-Identifier: Apache-2.0
-  --surface-secondary: var(--ant-color-bg-layout);
-  --surface-tertiary: var(--ant-color-fill-alter);
-  --surface-overlay: var(--ant-color-bg-mask);
+  --surface-primary: var(--ant-color-bg-container, #ffffff);
```
Adding default values effectively rolls back the use of `--ant-color-bg-container`, because now we still have to maintain one var definition for each theme, and it can get out of sync with AntD.

If there are some very specific use cases where variables defined on `.ant-layout` are not visible, then we can define colors just for those cases, without undermining the overall global approach for the majority of the components.

jkowall replied:
See comments below, if you want a follow up PR to fix these things I can work on implementing it.

jkowall commented:
## Response to Review Feedback

### @yurishkuro's comments:

**Q: What are the examples of 'fixed-position elements and components rendered outside the Layout'?**

The main example is `TimelineHeaderRow` which uses `position: fixed`. When an element is `position: fixed`, it's removed from the normal document flow and CSS custom properties defined on `.ant-layout` don't cascade to it - they need to be defined on an ancestor in the fixed element's inheritance chain (i.e., `body` or `:root`).

Additionally, several elements in the top navigation (`TopNav.css`, `ThemeToggleButton.css`, `TraceIDSearchInput.css`) use tokens like `--text-inverse` and they're rendered inside the Ant Design `<Header>` which may not always inherit from `.ant-layout` depending on the component tree.

**Re: fallback values undermining the global approach**

You raise a valid point. I tried to address this by removing the `body` selector and fallbacks, but it immediately broke the navigation bar because `--text-inverse` and other tokens weren't available.

**Proposed approach:**
1. Keep the current structure with `body, .ant-layout` for now to ensure all components work
2. The fallback values match Ant Design's default light theme, so they're effectively documentation of what Ant provides
3. In a follow-up PR, we could explore alternatives like:
   - Using `:root` instead of `body` for better inheritance
   - Ensuring Ant's `ConfigProvider` sets `cssVar` on `:root` so we don't need custom fallbacks

### @Copilot's comment on document.body null check:

✅ **Implemented** - Added the suggested null check to `getBackgroundColor()` in `render-into-canvas.tsx`:
```typescript
if (typeof document !== 'undefined' && document.body) {
  const isDark = document.body.getAttbute('data-theme') === 'dark';
  return isDark ? BG_COLOR_DARK : BG_COLOR_LIGHT;
}
return BG_COLOR_LIGHT;
```

This handles SSR and test environments gracefully.

jkowall commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/SpanGraph/render-into-canvas.tsx:
```diff
@@ -4,14 +4,27 @@
+  }
+  // Default to light background when document or document.body is unavailable (SSR, tests)
+  return BG_COLOR_LIGHT;
+}
```
Addressed - refactored renderIntoCanvas to accept isDark as a parameter, exported getBackgroundColor for testing, and added tests for both light and dark mode scenarios.

jkowall commented on packages/jaeger-ui/src/components/TracePage/TracePageHeader/SpanGraph/CanvasSpanGraph.tsx:
```diff
@@ -17,12 +18,13 @@ export const getColor = (hex: string) => colorGenerator.getRgbColorByKey(hex);
       renderIntoCanvas(canvasRef.current, items, valueWidth, getColor);
     }
-  }, [items, valueWidth]);
+  }, [items, valueWidth, mode]);
```
Addressed - refactored to pass isDark boolean from CanvasSpanGraph to renderIntoCanvas. Now there's a single source of truth (useThemeMode hook) and the function is more testable since it doesn't read from the DOM.
